File: leetcode/dp/tree/0124_binary_tree_max_path_sum.cpp
// LC 124. Binary Tree Maximum Path Sum | Hard (DP on Tree)
#include <algorithm>
#include <climits>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x) : val(x), left(nullptr), right(nullptr) {} };

class Solution {
    int ans = INT_MIN;
    int dfs(TreeNode* n) {
        if (!n) return 0;
        int l = max(dfs(n->left), 0), r = max(dfs(n->right), 0);
        ans = max(ans, n->val + l + r);
        return n->val + max(l, r);
    }
public:
    int maxPathSum(TreeNode* root) { dfs(root); return ans; }
};

File: leetcode/dp/tree/0124_binary_tree_max_path_sum.py
# LC 124. Binary Tree Maximum Path Sum | Hard (DP on Tree)
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right

class Solution:
    def maxPathSum(self, root):
        self.ans = float('-inf')
        def dfs(node):
            if not node: return 0
            l = max(dfs(node.left), 0); r = max(dfs(node.right), 0)
            self.ans = max(self.ans, node.val + l + r)
            return node.val + max(l, r)
        dfs(root); return self.ans
